common/credentials-interface/src/lib.rs

```diff
@@ -74,17 +90,168 @@ impl Display for CredentialType {
 
 #[derive(Debug, Clone)]
 pub struct CredentialSigningData {
-    pub pedersen_commitments_openings: Vec<Scalar>,
+    pub withdrawal_request: WithdrawalRequest,
 
-    pub blind_sign_request: BlindSignRequest,
+    pub request_info: RequestInfo,
 
-    pub public_attributes_plain: Vec<String>,
+    pub ecash_pub_key: PublicKeyUser,
+
+    pub expiration_date: u64,
 
     pub typ: CredentialType,
 }
 
-#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
 pub struct CredentialSpendingData {
+    pub payment: Payment,
+
+    pub pay_info: PayInfo,
+
+    pub spend_date: u64,
+
+    pub value: u64,
+
+    pub typ: CredentialType,
+
+    /// The (DKG) epoch id under which the credential has been issued so that the verifier could use correct verification key for validation.
+    pub epoch_id: u64,
+}
+
+impl CredentialSpendingData {
+    pub fn verify(
+        &self,
+        params: &Parameters,
+        verification_key: &VerificationKeyAuth,
+    ) -> Result<bool, CompactEcashError> {
+        self.payment.spend_verify(
+            params,
+            verification_key,
+            &self.pay_info,
+            date_scalar(self.spend_date),
+        )
+    }
+
+    pub fn serial_number_b58(&self) -> String {
+        self.payment.serial_number_bs58()
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        // simple length prefixed serialization
+        // TODO: change it to a standard format instead
+        let mut bytes = Vec::new();
+        let payment_bytes = self.payment.to_bytes();
+        let typ = self.typ.to_string();
+        let typ_bytes = typ.as_bytes();
+
+        bytes.extend_from_slice(&(payment_bytes.len() as u32).to_be_bytes());
+        bytes.extend_from_slice(&payment_bytes);
+        bytes.extend_from_slice(&self.pay_info.pay_info_bytes); //this is 72 bytes long
+        bytes.extend_from_slice(&self.spend_date.to_be_bytes());
+        bytes.extend_from_slice(&self.value.to_be_bytes());
+        bytes.extend_from_slice(&(typ_bytes.len() as u32).to_be_bytes());
+        bytes.extend_from_slice(typ_bytes);
+        bytes.extend_from_slice(&self.epoch_id.to_be_bytes());
+
+        bytes
+    }
+
+    pub fn try_from_bytes(raw: &[u8]) -> Result<Self, CompactEcashError> {
```

Comment:
I'm not going to try to validate your bounds here. I'm going to assume it's fine if you tested it to work